File: Binary Search/find-first-and-last-position-of-element-in-sorted-array.py
# 120ms 23.50%
# 14.3MB 54.02%
from typing import List

class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        length = len(nums)
        if length == 0:
            return [-1, -1]
        index1 = self.findMinimum(nums, target, 0, length-1)
        index2 = self.findMaximum(nums[index1:], target, 0, len(nums[index1:])-1)
        if index2 != -1:
            index2 += index1
        return [index1, index2]

    def findMinimum(self, nums, target, left, right):
        if left == right:
            if nums[right] == target:
                return right
            else:
                return -1
        k = int((left+right)/2)
        if nums[k] >= target:
            return self.findMinimum(nums, target, left, k)
        else:
            return self.findMinimum(nums, target, k+1, right)

    def findMaximum(self, nums, target, left, right):
        if left == right:
            if nums[left] == target:
                return left
            else:
                return -1
        k = int((left+right + 1)/2)
        if nums[k] > target:
            return self.findMaximum(nums, target, left, k-1)
        else:
            return self.findMaximum(nums, target, k, right)

l = [-5,-5,0,0,1,1,1,1,1,1,2,2,5,5,7,7]
l = [1,1,1,1,1,1,1,1,1]
a = Solution()
b = a.searchRange(l, 1)
print(b)

File: Array/find-first-and-last-position-of-element-in-sorted-array.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        // special cases
        if (nums.size() == 0) return vector<int>{-1, -1};
        if (nums.size() == 1) return nums[0] == target ? vector<int>{0, 0} : vector<int>{-1, -1};

        int left_anchor = 0, right_anchor = nums.size()-1;
        int left = 0, right = nums.size()-1;
        // left_anchor, left, right, right_anchor
        while (left_anchor <= right_anchor){
            int mid = left_anchor + (right_anchor - left_anchor) / 2;
            if (nums[mid] == target){
                left = mid;
                right = mid;
                break;
            }
            if (nums[mid] < target){
                left_anchor = mid + 1;
            } else{
                right_anchor = mid - 1;
            }
        }
        if (nums[left] != target) return vector<int>{-1, -1};

        while (left_anchor < left){
            int mid = left_anchor + (left - left_anchor) / 2;
            if (nums[mid] == target) left = mid;
            else left_anchor = mid + 1;
            }
        while (right < right_anchor){
            int mid = right + (right_anchor - right) / 2 + 1;
            if (nums[mid] == target) right = mid;
            else right_anchor = mid - 1;   
        }
        return vector<int>{left, right};
    }
};

int main() {
    Solution sol;
    vector<int> nums = {5,5,7,7,8,10,10};
    int target = 8;
    vector<int> ans = sol.searchRange(nums, target);
    for (int i = 0; i < ans.size(); i++){
        cout << ans[i] << " ";
    }
    cout << endl;
    return 0;
}